### DISCUSSION TITLE
Make working the EXE from terminal

### DISCUSSION THREAD

**Discussion by skillsfactory** [MatthiasValvekens/pyHanko#229] [Q&A]
Hello,

I made this code and using main.py work perfectly, but when I try to use it with builded exe I get this error:

ModuleNotFoundError: No module named 'pkcs11.defaults'

It seems that from the executable pyhanko ignores the PKCS11 module.
There's a way to make it work from terminal?
I've simply to sign a PDF in PAdES-BES

Here's the code:

`import os
import sys

from pyhanko.pdf_utils.incremental_writer import IncrementalPdfFileWriter
from pyhanko.sign import signers, pkcs11
from pyhanko.sign.fields import SigSeedSubFilter

from PyKCS11 import *

myfile = sys.argv[1]
mypin = sys.argv[2]

myfile_name, myfile_extension = os.path.splitext(myfile)
mysignedfile = myfile_name + 'signed' + myfile_extension
mylibpath = os.getenv('PYKCS11LIB')

mypkcs11 = PyKCS11Lib()
mypkcs11.load(mylibpath)

# get 1st slot
slot = mypkcs11.getSlotList(tokenPresent=True)[0]

tokenInfo = mypkcs11.getTokenInfo(slot)
mylabel = tokenInfo.label.strip()

sessiong = mypkcs11.openSession(slot, CKF_SERIAL_SESSION | CKF_RW_SESSION)
sessiong.login(mypin)

# find private key
myKeyID = ""

for myprkey in sessiong.findObjects([(CKA_CLASS, CKO_PRIVATE_KEY)]):

    if "DS" in sessiong.getAttributeValue(myprkey, [CKA_LABEL])[0]:
        myKeyID = sessiong.getAttributeValue(myprkey, [CKA_ID])[0]
        break

# find private cert
myCertID = ""

for mycid in sessiong.findObjects([(CKA_CLASS, CKO_CERTIFICATE)]):

    if "DS" in sessiong.getAttributeValue(mycid, [CKA_LABEL])[0]:
        myCertID = sessiong.getAttributeValue(mycid, [CKA_ID])[0]
        break

# logout
sessiong.logout()
sessiong.closeSession()

session=pkcs11.open_pkcs11_session(mylibpath, user_pin=mypin, token_label=mylabel)

signature_meta = signers.PdfSignatureMetadata(
    field_name='Signature', md_algorithm='sha256',
    # Mark the signature as a PAdES signature
    subfilter=SigSeedSubFilter.PADES,
    use_pades_lta=True
)

signer = pkcs11.PKCS11Signer(pkcs11_session=session, key_id=myKeyID, cert_id=myCertID,)

with open(myfile, 'rb') as inf:
    w = IncrementalPdfFileWriter(inf)
    with open(mysignedfile, 'wb') as outfile:
        signers.sign_pdf(w, signature_meta=signature_meta, signer=signer, output=outfile)`

Thanks everyone
Luca

**Comment by MatthiasValvekens** [comment]
Looks like you're trying to use pyHanko's built-in PKCS#11 integration (=`PKCS11Signer`) with PyKCS11. That won't work---the built-in integration uses `python-pkcs11` to talk to the token.

You have, broadly speaking, two options:

 - Replace PyKCS11 calls with corresponding ones from `python-pkcs11` in your code
 - Implement your own version of `PKCS11Signer` that uses PyKCS11 under the hood instead of `python-pkcs11`.

**Reply by skillsfactory** [reply]
But why from PyCharm run work perfectly?

**Reply by MatthiasValvekens** [reply]
Ah, sorry, I was a bit too quick to judge, you're only using PyKCS11 to run queries before starting the signing process (pro tip: use code blocks next time; easier on the eyes).

If it's not working when you bundle things as an executable, then there's likely an issue with some shared library or perhaps with the native part of `python-pkcs11`. I can't really help you with that; I've never used any of those "python to .exe" bundlers, nor do I have a Windows machine to test with. You should address your concerns to the `python-pkcs11` issue tracker, or (even better) take it up with whoever wrote the tool that you're using to build those .exe files.

**Reply by skillsfactory** [reply]
I solved, and also optimized the code, but in fact it was Pyinstaller's fault that omitted some libraries.
Anyway "I'm allergic to Windows" 😂; I've built an executable binary and work like a charm.
